File: Printly-backend/app/schemas/order.py
from pydantic import BaseModel, Field, ConfigDict
from uuid import UUID
from datetime import date, datetime
from decimal import Decimal

from app.enums import OrderStatus


class OrderItemCreate(BaseModel):
    book_id: UUID
    copies: int = Field(..., gt=0)


class OrderCreate(BaseModel):
    customer_id: UUID | None = Field(default=None)
    walk_in_customer_id: UUID | None = Field(default=None)
    notes: str | None = Field(default=None)
    due_date: date | None = Field(default=None)
    items: list[OrderItemCreate] = Field(..., min_length=1)


class OrderItemResponse(BaseModel):
    id: UUID
    book_id: UUID | None
    book_title: str
    copies: int
    unit_price: Decimal
    total_pages: int
    color_mode: str
    sides_per_page: int
    binding_type: str | None
    has_lamination: bool
    materials_snapshot: list[dict]
    subtotal: Decimal

    model_config = ConfigDict(from_attributes=True)


class OrderResponse(BaseModel):
    id: UUID
    order_number: str
    customer_id: UUID | None
    walk_in_customer_id: UUID | None
    created_by: UUID
    status: OrderStatus
    total_amount: Decimal
    paid_amount: Decimal
    notes: str | None
    due_date: date | None
    items: list[OrderItemResponse] = list()
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class OrderUpdate(BaseModel):
    due_date: date | None = Field(default=None)
    notes: str | None = Field(default=None)


class OrderStatusUpdate(BaseModel):
    status: OrderStatus


class OrdersRequest(BaseModel):
    status: OrderStatus | None = None
    customer_id: UUID | None = None
    walk_in_customer_id: UUID | None = None
    date_from: date | None = None
    date_to: date | None = None
    order_number: str | None = None
    offset: int = 0
    limit: int = 10
    order_by: str = "created_at"
    order_dir: str = "desc"


class OrdersListResponse(BaseModel):
    total: int
    orders: list[OrderResponse]


class OrdersCustomerRequest(BaseModel):
    status: OrderStatus | None = None
    date_from: date | None = None
    date_to: date | None = None
    offset: int = 0
    limit: int = 10
    order_by: str = "created_at"
    order_dir: str = "desc"

File: Printly-frontend/src/types/order.ts
// ==================== Order Types ====================

export type OrderStatus =
	| "new"
	| "printing"
	| "ready"
	| "delivered"
	| "cancelled";

export interface MaterialSnapshotItem {
	material_id: string;
	material_name: string;
	quantity_per_copy: number;
	price_per_unit: number;
}

export interface OrderItemResponse {
	id: string;
	book_id: string | null;
	book_title: string;
	copies: number;
	unit_price: number;
	total_pages: number;
	color_mode: string;
	sides_per_page: number;
	binding_type: string | null;
	has_lamination: boolean;
	materials_snapshot: MaterialSnapshotItem[];
	subtotal: number;
	created_at: string;
	updated_at: string;
}

export interface OrderItemCreate {
	book_id: string;
	copies: number;
}

export interface OrderResponse {
	id: string;
	order_number: string;
	customer_id: string | null;
	walk_in_customer_id: string | null;
	created_by: string;
	status: OrderStatus;
	total_amount: number;
	paid_amount: number;
	notes: string | null;
	due_date: string | null;
	items: OrderItemResponse[];
	created_at: string;
	updated_at: string;
}

export interface OrderCreate {
	customer_id?: string;
	walk_in_customer_id?: string;
	notes?: string;
	due_date?: string;
	items: OrderItemCreate[];
	paid_amount?: number;
}

export interface OrderUpdate {
	due_date?: string;
	notes?: string;
}

export interface OrderStatusUpdate {
	status: OrderStatus;
}

export interface OrdersRequest {
	status?: OrderStatus;
	customer_id?: string;
	walk_in_customer_id?: string;
	date_from?: string;
	date_to?: string;
	order_number?: string;
	offset?: number;
	limit?: number;
}

export interface OrdersListResponse {
	total: number;
	orders: OrderResponse[];
}
